src/linear_algebra/finsupp.lean

```diff
@@ -141,247 +97,320 @@ begin
 
-lemma equiv_of_dim_eq_dim (h : dim α β = dim α γ) : nonempty (β ≃ₗ[α] γ) :=
+theorem supported_Union {δ : Type*} (s : δ → set α) :
+  supported β γ (⋃ i, s i) = ⨆ i, supported β γ (s i) :=
 begin
-  rcases exists_is_basis α β with ⟨b, hb⟩,
-  rcases exists_is_basis α γ with ⟨c, hc⟩,
-  rw [← hb.mk_eq_dim, ← hc.mk_eq_dim] at h,
-  rcases quotient.exact h with ⟨e⟩,
-  exact ⟨ (module_equiv_lc hb).trans (linear_equiv.trans (lc.congr α b c e) (module_equiv_lc hc).symm) ⟩
+  refine le_antisymm _ (supr_le $ λ i, supported_mono $ set.subset_Union _ _),
+  haveI := classical.dec_pred (λ x, x ∈ (⋃ i, s i)),
+  suffices : ((submodule.subtype _).comp (restrict_dom β γ (⋃ i, s i))).range ≤ ⨆ i, supported β γ (s i),
+  { rwa [linear_map.range_comp, range_restrict_dom, map_top, range_subtype] at this },
+  rw [range_le_iff_comap, eq_top_iff],
+  rintro l ⟨⟩, rw mem_coe,
+  apply finsupp.induction l, {exact zero_mem _},
+  refine λ x a l hl a0, add_mem _ _,
+  haveI := classical.dec_pred (λ x, ∃ i, x ∈ s i),
+  by_cases (∃ i, x ∈ s i); simp [h],
+  { cases h with i hi,
+    exact le_supr (λ i, supported β γ (s i)) i (single_mem_supported γ _ hi) },
+  { rw filter_single_of_neg,
+    { simp },
+    { exact h } }
 end
 
-lemma eq_bot_iff_dim_eq_zero (p : submodule α β) (h : dim α p = 0) : p = ⊥ :=
+theorem supported_union (s t : set α) :
+  supported β γ (s ∪ t) = supported β γ s ⊔ supported β γ t :=
+by erw [set.union_eq_Union, supported_Union, supr_bool_eq]; refl
+
+theorem supported_Inter {ι : Type*} (s : ι → set α) :
+  supported β γ (⋂ i, s i) = ⨅ i, supported β γ (s i) :=
 begin
-  have : dim α p = dim α (⊥ : submodule α β) := by rwa [dim_bot],
-  rcases equiv_of_dim_eq_dim this with ⟨e⟩,
-  exact e.eq_bot_of_equiv _
+  refine le_antisymm (le_infi $ λ i, supported_mono $ set.Inter_subset _ _) _,
+  simp [le_def, infi_coe, set.subset_def],
+  exact λ l, set.subset_Inter
 end
 
-lemma injective_of_surjective (f : β →ₗ[α] γ)
-  (hβ : dim α β < cardinal.omega) (heq : dim α γ = dim α β) (hf : f.range = ⊤) : f.ker = ⊥ :=
-have hk : dim α f.ker < cardinal.omega := lt_of_le_of_lt (dim_submodule_le _) hβ,
+set_option class.instance_max_depth 37
+def supported_equiv_finsupp (s : set α) [decidable_pred (λ (x : α), x ∈ s)] :
+  (supported β γ s) ≃ₗ[γ] (s →₀ β) :=
+(restrict_support_equiv s).to_linear_equiv
 begin
-  rcases cardinal.lt_omega.1 hβ with ⟨d₁, eq₁⟩,
-  rcases cardinal.lt_omega.1 hk with ⟨d₂, eq₂⟩,
-  have : 0 = d₂,
-  { have := dim_eq_surjective f (linear_map.range_eq_top.1 hf),
-    rw [heq, eq₁, eq₂, ← nat.cast_add, cardinal.nat_cast_inj] at this,
-    exact nat.add_left_cancel this },
-  refine eq_bot_iff_dim_eq_zero _ _,
-  rw [eq₂, ← this, nat.cast_zero]
+  show is_linear_map γ ((lsubtype_domain s : (α →₀ β) →ₗ[γ] (s →₀ β)).comp
+      (submodule.subtype (supported β γ s))),
+  exact linear_map.is_linear _
 end
+set_option class.instance_max_depth 32
+
+def lsum [decidable_eq γ] (f : α → γ →ₗ[γ] β) : α →₀ γ →ₗ[γ] β :=
```

Comment:
The precedence of the arrows here is wrong. I think all the arrows should be at the same precedence level (=25)

Reply:
I agree, but I don't know why this happens. Could it be in the declaration
```
notation β ` →ₗ[`:25 α `] ` γ := linear_map α β γ
```
The 25 seems to refer only to the first part of the notation. Could that be the reason?

Reply:
Yes. You can set the binder power of all tokens and all variables (except maybe the first?)
I think the best way to do it is:
```
notation β ` →ₗ[`:25 α:0 `] `:0 γ:25 := linear_map α β γ
```
At least, that was the way it was done for similar notation back in Lean 2 (written by Leo):
https://github.com/leanprover/lean2/blob/master/hott/init/reserved_notation.hlean#L230

Reply:
That doesn't seem to do the trick: `α →₀ γ →ₗ[γ] β` is still parsed as `(α →₀ γ) →ₗ[γ] β`. I don't really understand what these numbers mean. Any other suggestions?

Reply:
Ok, Floris, you were right, but the power of `→₀` wasn't defined correctly either: It said
```
reserve infix ` →₀ `:25
[...]
infixr →₀ := finsupp
```
Apparently, the 25 didn't come through here. If I just write
```
[...]
infixr ` →₀ `:25 := finsupp
```
instead, it works.